src/skmultiflow/metrics/_classification_performance_evaluator.pyx

```diff
@@ -0,0 +1,918 @@
+        cdef double sum_f1
+        if n_classes == 2:
+            # Binary case
+            return self._f1_per_class(1)
+        if class_value == -1 and class_value == -1:
+            # Multi-class case, calculate 'macro' average
+            sum_f1 = 0.0
+            for i in self.confusion_matrix.classes:
+                sum_f1 += self._f1_per_class(i)
+            return sum_f1 / <double>n_classes
+        else:
+            # Calculate for specified class
+            return self._f1_per_class(class_value)
+
+    cdef _f1_per_class(self, int class_value):
+        cdef double precision
+        cdef double recall
+        if self.confusion_matrix.n_classes > 0:
+            precision = self.precision_score(class_value)
+            recall = self.recall_score(class_value)
+            # if ill-defined, return 0.0
+            return (2.0 * (precision * recall) / (precision + recall)) if (precision + recall) > 0.0 else 0.0
+        return 0.0
+
+    cpdef double geometric_mean_score(self):
+        """ Geometric mean score.
+        
+        The geometric mean is a good indicator of a classifier's performance
+        in the presence of class imbalance because it is independent of the
+        distribution of examples between classes [1]_. This implementation
+        computes the geometric mean of class-wise sensitivity (recall)
+        
+        .. math::
+            gm = \sqrt[n]{s_1\cdot s_2\cdot s_3\cdot \ldots\cdot s_n}
+        
+        where :math:`s_i` is the sensitivity (recall) of class :math:`i` and : math: `n`
+        is the number of classes.
+        
+        Returns
+        -------
+        float
+            Geometric mean score.
+        
+        References
+        ----------
+        .. [1] Barandela, R. et al. “Strategies for learning in class imbalance problems”,
+               Pattern Recognition, 36(3), (2003), pp 849-851.
+        
+        """
+        cdef int n_classes = self.confusion_matrix.n_classes
+        cdef np.ndarray[DTYPE_t, ndim=1] sensitivity_per_class = np.zeros(n_classes, dtype=DTYPE)
+        cdef unsigned int i, c
+        if n_classes > 0:
+            for i, c in enumerate(self.confusion_matrix.classes):
+                sensitivity_per_class[i] = self._recall_per_class(c)
+            with np.errstate(divide='ignore', invalid='ignore'):
+                return <double>sp.stats.gmean(sensitivity_per_class)
+        return 0.0
+
+    # cdef double _specificity_per_class(self, int class_value):
+    #     cdef double tn
+    #     cdef double fp
+    #     if self.confusion_matrix.n_classes > 0:
+    #         tn = self.confusion_matrix.sum_diag - self.confusion_matrix[(class_value, class_value)]
+    #         fp = self.confusion_matrix.sum_col[class_value] - self.confusion_matrix[(class_value, class_value)]
+    #         # if ill-defined, return 0.0
+    #         return tn / (tn + fp) if (tn + fp) > 0.0 else 0.0
+    #     return 0.0
```

Comment:
Is there any future purpose for this commented portion?

Reply:
This was used for the calculation of geometric mean for binary classification. It is not required in the new implementation. However, it might be handy in the future. I will add a comment about it.